### PR TITLE
chore: convert tools/config folder to TS

### PR DESCRIPTION
**Pull request by troZee**
Part of https://github.com/react-native-community/cli/issues/683#issuecomment-529863615 

### Current progress

- [x] findAssets-test
- [x] findDependencies-test
- [x] index-test
- [x] errors
- [x] findAssets
- [x] findDependencies
- [x] index
- [x] readConfigFromDisk
- [x] resolveNodeModuleDir
- [x] resolveReactNativePath
- [x] schema

### Problem description:

File `readConfigFromDisk.js` contains methods, which use cosmiconfig dependency or just convert `package.json` file. I have a problem to type this readers correctly and map returned values into `UserConfig` or `UserDependencyConfig`. All methods from `readConfigFromDisk` file should return `UserConfig` or `UserDependencyConfig`(this type should be created based on schema validator).

### CODE REVIEW

File: packages/cli-types/src/index.ts [hunk 1]
```diff
@@ -134,6 +134,89 @@ export interface Config {
...
+    [name: string]: PlatformConfig<any, any, any, any>;
+  } | null;
+  commands?: Command[];
+  haste?: {
+    platforms: Array<string>;
+    providesModuleNodeModules: Array<string>;
+  } | null;
+};
+
+// export type UserDependencyConfig = {
```
Comment (ptrocki) [inline packages/cli-types/src/index.ts]: Will be deleted

File: packages/cli/src/tools/config/readConfigFromDisk.ts [hunk 2]
```diff
@@ -1,21 +1,17 @@
-/**
```
Comment (ptrocki) [inline packages/cli/src/tools/config/readConfigFromDisk.ts]: Need help here

**Comment by nickytonline** [comment]
> 
> ### Problem description:
> File `readConfigFromDisk.js` contains methods, which use cosmiconfig dependency or just convert `package.json` file. I have a problem to type this readers correctly and map returned values into `UserConfig` or `UserDependencyConfig`. All methods from `readConfigFromDisk` file should return `UserConfig` or `UserDependencyConfig`(this type should be created based on schema validator).


It's not clear to me what you're having a hard time typing. Is it migrating the type from Flow to TypeScript that is the issue for you?

I [created a TypeScript playground](https://www.typescriptlang.org/play/#code/PQKgBAKg8gIlDKYTAFAEsB2AXApgJwDMBDAYxzAAU8B7AKxxKwqLyIFsBnAQQwBMa0vMAG8Avumz5iZSjXqNmrTgEkEI8aEiw1yFBNyFS5CgBsiWAtTxsAwtQwE0AcwA8KMLLoMmLdhwA07mAwOAAOOHwRJACein6BHlRejHYOzgnBYRG8UdGpjk6BAHzqepgG0uT5zhAiQcDAYABK1NRYYADuABb45Fg9YCT2BQCurFho9mBdRBxgAEY4EWB4OBzUJgBuOEIENGxBNG0AXGAcWHiYTgDceh4NlOZdYFjUzTik7QBy5mjbZ9QxmRDh9GD8JttmP1TudLhgbncwA8AHSosBsKzkUI0cJ4CZrREPAASs1wg2GzjGvymq3WWx2C1mDKmRE2RDQZnmJixJhGTkwHCCM3OOFOwiCHlCZgsVk4py4eFY0RcsKuRQykpom0EawAstReCNuV8DTh9YbuRx5YqiMrVfD1UFRLdxCgHgB1IjYF5vahsNDtRYkIgjDjkTBgIhgUP4clpJxgABEq0+4L+OChXUTYDQc2ooQm9iIJn8SeFuGzXqEydaWErq2jGAA1hhqB0MGAABSkIZ4XhXH3kthsCJYOYRoY5MCObkASjdjSDIbDLx60UjDeDJhIRvMDKrgy9rcD5ByjgwO2RC4XYAAYlYwFB-VgXBBSwBpdVnJbTLBYUJWg0OSbFy1BOBwyL+iQNDrAQWDIkMbDAFg0ThBw0FoAWwBHoCGAkFcAC0KFoRhBYEQAzARACswAAMT9DgBF+gGBE9CYuJEahOD6FIRhgAAqmGeDVE4tQ4AAHrgfBzE+AYuLJL4KS4IlvkmKZgr8kJPImX6JuWOA6aWNZtDpdT3I02I4AQsYjM2rbtmA1DbHgkYYOu6yruYYDRICG7kJYeBkEIrxgLiAVsJGLxcYMPQkE2CxWZi0YcAOAa-v+gHAGwLBoKG6FdEaABeCF+sAoFOMhPQEbZLZthgnHhARmANWspFYEKpKio2tXtrcHhHFgpw1fZGB9SsoJYGmWn9AA-DCFxXLc9SNFwvD9oWGDFqFcjeN+f5XIKmrJINZkeK5-DUIIc2ePIPhKNwfACLwGo5tQHDXUkt1xCoCAvQA2k2ODRPNcJOAAukNdl1RkzooKIQA) for getting started with the `ConfigT` and `UserConfigT` conversion from Flow.

I also came across this utility tonight which you may find helpful, care of Kahn Academy, https://flow-to-ts.netlify.com.

**Comment by thymikee** [comment]
@ptrocki I've pushed a bunch of adjustments to unblock you. This looks mostly good now, but I'd like you and @Esemesek to check it once again.

**Comment by ptrocki** [comment]
Tested below commands:  
```
node ./cli/packages/cli/build/index.js init AwesomeCLI 
cd AwesomeCLI
yarn add -D @react-native-community/netinfo
cd ios && pod install && cd ..
node ../cli/packages/cli/build/index.js run-ios
node ../cli/packages/cli/build/index.js run-android
node ../cli/packages/cli/build/index.js config  
```
and it works for both platforms. In this case netinfo was added to project correctly.

File: packages/cli-types/src/index.ts [hunk 3]
```diff
@@ -134,17 +136,15 @@ export interface Config {
...
 
 /**
- * Shares some structure with ConfigT, except that haste and root
+ * Shares some structure with Config, except that haste and root
  * are calculated and can't be defined
  */
-export type UserConfig = Diff<Config, {root: any; haste: any}> & {
-  reactNativePath: string | void;
 
+export type UserConfig = Omit<Omit<Config, 'root'>, 'haste'> & {
```
Comment (Esemesek) [inline packages/cli-types/src/index.ts]: I think this can be replaced by: `Omit<Config, 'root' | 'haste'>`

File: packages/cli-types/src/index.ts [hunk 4]
```diff
@@ -153,6 +153,20 @@ export type UserConfig = Diff<Config, {root: any; haste: any}> & {
   };
 };
 
+export type UserDependencyConfig = {
+  // Additional dependency settings
+  dependency: Omit<Omit<Dependency, 'name'>, 'root'>;
```
Comment (Esemesek) [inline packages/cli-types/src/index.ts]: Same as above

File: packages/cli-types/src/index.ts (hunk 3 above)
Comment (nickytonline) [reply packages/cli-types/src/index.ts]: Correct. When I gave him the example, I typed it like that by accident. My bad 😅. I updated the playground example up top the other week to reflect that.

Comment (thymikee) [reply packages/cli-types/src/index.ts]: My bad, fixed :)

**Pull request merged**